File: 01_language_comparison/graph_algorithm/cycle_detection.rb
class Graph
    attr_accessor :vertices, :adjacency_list
    
    def initialize(vertices)
      @vertices = vertices
      @adjacency_list = Array.new(vertices) { [] }
    end
    
    def add_edge(source, destination)
      @adjacency_list[source].push(destination)
    end
    
    def has_cycle?
      visited = Array.new(@vertices, false)
      rec_stack = Array.new(@vertices, false)
      
      @vertices.times do |i|
        if !visited[i]
          if cycle_util(i, visited, rec_stack)
            return true
          end
        end
      end
      
      return false
    end
    
    private
    
    def cycle_util(vertex, visited, rec_stack)
      visited[vertex] = true
      rec_stack[vertex] = true
      
      @adjacency_list[vertex].each do |neighbor|
        if !visited[neighbor]
          if cycle_util(neighbor, visited, rec_stack)
            return true
          end
        elsif rec_stack[neighbor]
          return true
        end
      end
      
      rec_stack[vertex] = false
      return false
    end
  end
  
  # Example usage
  def run_cycle_detection
    start_time = Time.now
    
    # Create a graph with 4 vertices
    g = Graph.new(4)
    g.add_edge(0, 1)
    g.add_edge(0, 2)
    g.add_edge(1, 2)
    g.add_edge(2, 0)
    g.add_edge(2, 3)
    g.add_edge(3, 3)
    
    if g.has_cycle?
      puts "Graph contains cycle"
    else
      puts "Graph doesn't contain cycle"
    end
    
    end_time = Time.now
    puts "Execution time: #{(end_time - start_time) * 1000} ms"
  end
  
  run_cycle_detection

File: 01_language_comparison/graph_algorithm/cycle_detection.go
package main

import (
	"fmt"
	"time"
)

// Graph represents a directed graph using adjacency list
type Graph struct {
	vertices int
	adjList  [][]int
}

// NewGraph creates a new graph with v vertices
func NewGraph(v int) *Graph {
	adjList := make([][]int, v)
	for i := range adjList {
		adjList[i] = make([]int, 0)
	}
	return &Graph{
		vertices: v,
		adjList:  adjList,
	}
}

// AddEdge adds an edge from src to dest
func (g *Graph) AddEdge(src, dest int) {
	g.adjList[src] = append(g.adjList[src], dest)
}

// IsCyclicUtil is a recursive helper function for cycle detection
func (g *Graph) IsCyclicUtil(v int, visited, recStack []bool) bool {
	visited[v] = true
	recStack[v] = true

	// Visit all the neighbors
	for _, neighbor := range g.adjList[v] {
		// If neighbor is not visited, check if there's a cycle starting from neighbor
		if !visited[neighbor] {
			if g.IsCyclicUtil(neighbor, visited, recStack) {
				return true
			}
		} else if recStack[neighbor] { // If neighbor is in recursion stack, cycle found
			return true
		}
	}

	// Remove the vertex from recursion stack before returning
	recStack[v] = false
	return false
}

// HasCycle detects if there is a cycle in the directed graph
func (g *Graph) HasCycle() bool {
	visited := make([]bool, g.vertices)
	recStack := make([]bool, g.vertices)

	// Call the recursive helper function for all vertices
	for i := 0; i < g.vertices; i++ {
		if !visited[i] {
			if g.IsCyclicUtil(i, visited, recStack) {
				return true
			}
		}
	}
	return false
}

func main() {
	startTime := time.Now()

	// Create a graph with 4 vertices
	g := NewGraph(4)
	g.AddEdge(0, 1)
	g.AddEdge(0, 2)
	g.AddEdge(1, 2)
	g.AddEdge(2, 0)
	g.AddEdge(2, 3)
	g.AddEdge(3, 3)

	if g.HasCycle() {
		fmt.Println("Graph contains cycle")
	} else {
		fmt.Println("Graph doesn't contain cycle")
	}

	executionTime := time.Since(startTime).Milliseconds()
	fmt.Printf("Execution time: %d ms\n", executionTime)
}